snooty/semanticparser.py

```diff
@@ -37,67 +43,42 @@ def run_event_parser(
     ) -> Dict[str, SerializableType]:
         event_parser = EventParser()
-
-        # Add event listeners here before calling consume()
-
+        event_parser.add_event_listener(
+            OBJECT_START_EVENT, self.build_slug_title_mapping
+        )
         event_parser.consume(pages)
 
         # Return dict containing fields updated in event-based parse
-        return {}
+        return {"slugToTitle": self.slug_title_mapping}
 
     # Returns a list of transformations to include in self.run()
     def functions(
         self, fn_names: List[str]
     ) -> List[Callable[[Dict[FileId, Page]], Dict[str, SerializableType]]]:
         fn_mapping = {
             "toctree": self.build_toctree,
-            "slug-title": self.build_slug_title,
             "breadcrumbs": self.breadcrumbs,
             "toctree order": self.toctree_order,
         }
 
         return [fn_mapping[name] for name in fn_names]
 
-    def build_slug_title(
-        self, pages: Dict[FileId, Page]
-    ) -> Dict[str, SerializableType]:
-        # Function which returns a dictionary of slug-title mappings
-        slug_title_dict: Dict[str, SerializableType] = {}
-        for file_id, page in pages.items():
-            # remove the file extension from the slug
-            slug = file_id.without_known_suffix
-
-            # Skip slug-title mapping if the file is an `includes`
-            if "includes" in slug or "images" in slug:
-                continue
-
-            # Parse for title
-            title = ""
-            ast: Dict[str, Any] = cast(Dict[str, Any], page.ast)
-            title_is_set = False
-
-            if ast is None:
-                return {}
-
-            for child in ast["children"]:
-                if title_is_set:
-                    break
-                if child["type"] == "section":
-                    for section_child in child["children"]:
-                        if title_is_set:
-                            break
-                        if section_child["type"] == "heading":
-                            for heading_child in section_child["children"]:
-                                if title_is_set:
-                                    break
-                                if heading_child["type"] == "text":
-                                    title = heading_child["value"]
-                                    title_is_set = True
-            slug_title_dict[slug] = title
-
-        self.slug_title = {"slugToTitle": slug_title_dict}
-
-        return self.slug_title
+    def build_slug_title_mapping(
+        self, filename: FileId, *args: SerializableType, **kwargs: SerializableType
+    ) -> None:
+        """Construct a slug-title mapping of all pages in property"""
+        obj = cast(Dict[str, SerializableType], kwargs.get("obj"))
+        slug = filename.without_known_suffix
+
+        # Don't parse these files for headings
+        if "includes" in slug or "images" in slug:
```

Comment:
Don't we just want to scan `.txt` files, allowing us to ask `if filename.suffix == ".txt"`?

If this is really what we want, it may still be more broad than we'd like: perhaps `if "includes" in filename.parts or "images" in filename.parts`?

Reply:
Yes you're right! The first bit is exactly what we want, thanks 👍